core/src/test/java/org/apache/iceberg/RandomVariants.java

```diff
@@ -0,0 +1,127 @@
+        randomNames.add(RandomUtil.generateString(10, random));
+      }
+      metadata =
+          VariantMetadata.from(VariantTestUtil.createMetadata(randomNames, random.nextBoolean()));
+    } else {
+      metadata = VariantMetadata.empty();
+    }
+
+    return Variant.of(metadata, randomVariant(random, metadata, randomType(random)));
+  }
+
+  private static PhysicalType randomType(Random random) {
+    PhysicalType[] types = PhysicalType.values();
+    return types[random.nextInt(types.length)];
+  }
+
+  private static VariantValue randomVariant(
+      Random random, VariantMetadata metadata, PhysicalType type) {
+    switch (type) {
+      case NULL:
+        return Variants.ofNull();
+      case BOOLEAN_TRUE:
+        return Variants.of(true);
+      case BOOLEAN_FALSE:
+        return Variants.of(false);
+      case INT8:
+        return Variants.of(
+            type,
+            ((Integer) RandomUtil.generatePrimitive(Types.IntegerType.get(), random)).byteValue());
+      case INT16:
+        return Variants.of(
+            type,
+            ((Integer) RandomUtil.generatePrimitive(Types.IntegerType.get(), random)).shortValue());
+      case INT32:
+        return Variants.of(type, RandomUtil.generatePrimitive(Types.IntegerType.get(), random));
+      case INT64:
+        return Variants.of(type, RandomUtil.generatePrimitive(Types.LongType.get(), random));
+      case DOUBLE:
+        return Variants.of(type, RandomUtil.generatePrimitive(Types.DoubleType.get(), random));
+      case DECIMAL4:
+        return Variants.of(
+            type, RandomUtil.generatePrimitive(Types.DecimalType.of(9, random.nextInt(9)), random));
+      case DECIMAL8:
+        return Variants.of(
+            type,
+            RandomUtil.generatePrimitive(Types.DecimalType.of(18, random.nextInt(18)), random));
+      case DECIMAL16:
+        return Variants.of(
+            type,
+            RandomUtil.generatePrimitive(Types.DecimalType.of(38, random.nextInt(38)), random));
+      case DATE:
+        return Variants.of(type, RandomUtil.generatePrimitive(Types.DateType.get(), random));
+      case TIMESTAMPTZ:
+        return Variants.of(
+            type, RandomUtil.generatePrimitive(Types.TimestampType.withZone(), random));
+      case TIMESTAMPNTZ:
+        return Variants.of(
+            type, RandomUtil.generatePrimitive(Types.TimestampType.withoutZone(), random));
+      case FLOAT:
+        return Variants.of(type, RandomUtil.generatePrimitive(Types.FloatType.get(), random));
+      case BINARY:
+        byte[] randomBytes = (byte[]) RandomUtil.generatePrimitive(Types.BinaryType.get(), random);
+        return Variants.of(type, ByteBuffer.wrap(randomBytes));
+      case STRING:
+        return Variants.of(type, RandomUtil.generatePrimitive(Types.StringType.get(), random));
+      case ARRAY:
+        // for now, generate an object instead of an array
+      case OBJECT:
+        ShreddedObject object = Variants.object(metadata);
```

Comment:
Seems we only have fully shredded objects here. Should we add partial shredded objects?

Reply:
This produces a variant in memory, not a serialized variant. The writer is free to serialize however it chooses, although this is currently not intended to exercise shredding paths that need records to be make sense.